### ISSUE TITLE
vmap over complex FlatMapping without specifying structure explicitly

### ISSUE THREAD

**Issue by frechette-alex** [google/jax#4132]
I have a complex [`FlatMapping`](https://github.com/deepmind/dm-haiku/blob/master/haiku/_src/data_structures.py#L134) with multiple layers (1). I'd like to `vmap` over that `FlatMapping`, but without being explicit about the structure of the `FlatMapping` : the structure may change (2), but I know I'll always be able to map over the leaves' first axis. Is this possible? How should I specify the `in_axes`? 

---

(1) The complex `FlatMapping` is the Jacobian of a neural net's lost function. Here is its `jax.tree_structure`:
```
PyTreeDef(haiku._src.data_structures.FlatMapping[PyTreeDef(dict[['EnvironmentModel/~/head_discount', 'EnvironmentModel/~/head_reward', 'EnvironmentModel/~/head_step_type', 'EnvironmentModel/~/linear', 'EnvironmentModel/~/torso/~/linear_0', 'EnvironmentModel/~/torso/~/linear_1']], [PyTreeDef(haiku._src.data_structures.FlatMapping[PyTreeDef(dict[['b', 'w']], [*,*])], [*,*]),PyTreeDef(haiku._src.data_structures.FlatMapping[PyTreeDef(dict[['b', 'w']], [*,*])], [*,*]),PyTreeDef(haiku._src.data_structures.FlatMapping[PyTreeDef(dict[['b', 'w']], [*,*])], [*,*]),PyTreeDef(haiku._src.data_structures.FlatMapping[PyTreeDef(dict[['b', 'w']], [*,*])], [*,*]),PyTreeDef(haiku._src.data_structures.FlatMapping[PyTreeDef(dict[['b', 'w']], [*,*])], [*,*]),PyTreeDef(haiku._src.data_structures.FlatMapping[PyTreeDef(dict[['b', 'w']], [*,*])], [*,*])])], [*,*,*,*,*,*,*,*,*,*,*,*])
```

(2) As the `FlatMapping` is a representation of a neural net's parameters, if the neural net architecture change, the `FlatMapping`'s dict structure will also change.

**Comment by tomhennigan** [comment]
If your function is `f(params, x)` you can use `jax.vmap(f, in_axes=(0, None))` to tell JAX you want to map the entire params dict (`in_axes` takes positional arguments by index, so `0` would refer to the entire tree in the 0th positional argument). The `None` for the second argument can obviously be changed as you like 😄

**Comment by frechette-alex** [comment]
If I simply have
```
def f(grad):
  print(jax.tree_map(lambda leaf: leaf.shape, grad))
```
and do `jax.vmap(f)(jac)`, the function `f` is only called once, but it does return the right shapes for the leaf nodes (first axis gone). Shouldn't the function be called multiple times?

I'd like to use each row of the Jacobian to do a gradient-based parameter update. Right now, I am doing:
```
for output_index in range(4):
     grad = jax.tree_map(lambda leaf: leaf[output_index], jac)
     updates, opt_state = self._optimizer.update(grad, opt_state)
     params = optax.apply_updates(params, updates)
```
The end goal is to try to use `vmap` to replace some of the logic above.

**Comment by tomhennigan** [comment]
Your example looks almost like a `scan` to me (loop and accumulate) rather than a `map` (parallel execute) since you are updating the params and opt_state at each iteration (you are "mapping" your grads and "carrying" your params/opt_state).

Would suggest looking at https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html

Re calling your function multiple times, the way `vmap` is implemented is that it traces your function by calling it once, and replaces all jax primitives in your function with versions that support inputs with an additional batch dimension. As such JAX now has a version of your function that can efficiently be applied in parallel on batches of data.

**Comment by frechette-alex** [comment]
Thanks for the resource, `scan` is the way to go.
